packages/base/matrix-room.gts

```diff
@@ -0,0 +1,415 @@
+  static isolated = class Isolated extends RoomMemberView {};
+  // The edit template is meant to be read-only, this field card is not mutable
+  static edit = class Edit extends RoomMemberView {};
+}
+
+class ScrollIntoView extends Modifier {
+  modify(element: HTMLElement) {
+    element.scrollIntoView();
+  }
+}
+
+class MessageCard extends Card {
+  @field author = contains(RoomMemberCard);
+  @field message = contains(MarkdownCard);
+  @field formattedMessage = contains(StringCard);
+  @field created = contains(DateTimeCard);
+  @field attachedCard = contains(Card);
+  @field index = contains(IntegerCard);
+
+  static embedded = class Embedded extends Component<typeof this> {
+    // TODO need to add the message specific CSS here
+    <template>
+      <BoxelMessage
+        {{ScrollIntoView}}
+        data-test-message-idx={{@model.index}}
+        data-test-message-card={{@model.attachedCard.id}}
+        @name={{@model.author.displayName}}
+        @datetime={{formatRFC3339 this.timestamp}}
+        style={{cssVar
+          boxel-message-avatar-size=messageStyle.boxelMessageAvatarSize
+          boxel-message-meta-height=messageStyle.boxelMessageMetaHeight
+          boxel-message-gap=messageStyle.boxelMessageGap
+          boxel-message-margin-left=messageStyle.boxelMessageMarginLeft
+        }}
+      >
+        {{! template-lint-disable no-triple-curlies }}
+        {{{@model.formattedMessage}}}
+
+        {{#if @model.attachedCard}}
+          <this.cardComponent />
+        {{/if}}
+      </BoxelMessage>
+    </template>
+
+    get timestamp() {
+      if (!this.args.model.created) {
+        throw new Error(`message created time is undefined`);
+      }
+      return this.args.model.created.getTime();
+    }
+
+    get cardComponent() {
+      if (!this.args.model.attachedCard) {
+        return;
+      }
+      return this.args.model.attachedCard.constructor.getComponent(
+        this.args.model.attachedCard,
+        'isolated'
+      );
+    }
+  };
+  // The edit template is meant to be read-only, this field card is not mutable
+  static edit = class Edit extends JSONView {};
+}
+
+export class MatrixRoomCard extends Card {
+  // the only writeable field for this card should be the "events" field.
+  // All other fields should derive from the "events" field.
+  @field events = containsMany(MatrixEventCard);
+
+  @field roomId = contains(StringCard, {
+    computeVia: function (this: MatrixRoomCard) {
+      return this.events.length > 0 ? this.events[0].room_id : undefined;
+    },
+  });
+
+  @field name = contains(StringCard, {
+    computeVia: function (this: MatrixRoomCard) {
+      let events = this.events
+        .filter((e) => e.type === 'm.room.name')
+        .sort((a, b) => a.origin_server_ts - b.origin_server_ts) as
+        | RoomNameEvent[];
+      if (events.length > 0) {
+        return events.pop()!.content.name;
+      }
+      return; // this should never happen
+    },
+  });
```

Comment:
Is it a concern that we are re-evaluating this and other computed every time an event comes in when it is very unlikely they will be changing?

Reply:
yeah i'm a bit concerned about that as well. it would be nice if we had some kind of memoization features in cards (like `@cached` in ember. we could try memoizing card state in module scope since the room cards are special in that they are session scoped on the host... probably we'd want to talk with @ef4 about that first though...

Reply:
I can prototype something that I think could work.